File: src/main/java/nnglebanov/daoexample/domain/Comment.java
package nnglebanov.daoexample.domain;

import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import javax.persistence.*;
import java.util.Date;


@Data
@EqualsAndHashCode(exclude = "book")
@ToString(exclude = "book")
@AllArgsConstructor
@RequiredArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "COMMENTS")
public class Comment {
    @ManyToOne
    Book book;
    @NonNull
    private String commentText;
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id = 0;
    @CreationTimestamp
    private Date createdAt;

    @PreRemove
    private void removeComment() {
        book.getComments().remove(this);
    }
}

File: src/main/java/nnglebanov/daoexample/domain/Comment.kt
package nnglebanov.daoexample.domain

import nnglebanov.daoexample.helpers.Utils
import org.springframework.data.mongodb.core.mapping.DBRef
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.Field
import org.springframework.data.mongodb.core.mapping.MongoId
import java.util.*

@Document
data class Comment(
        @DBRef val book: Book?,
        @Field("message") val message: String,
        @Field("dateTime") val dateTime: Date
) {
    @MongoId
    var id: String = Utils.generateId()
}
